maven-plugin/src/main/java/io/confluent/kafka/schemaregistry/maven/derive/schema/DeriveJsonSchema.java

```diff
@@ -0,0 +1,109 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.kafka.schemaregistry.maven.derive.schema;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.annotations.VisibleForTesting;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.stream.Collectors;
+
+public class DeriveJsonSchema extends DeriveSchema {
+
+  public static final String NUMBER = "number";
+  public static final String STRING = "string";
+  public static final String BOOLEAN = "boolean";
+  public static final String NULL = "null";
+
+  public DeriveJsonSchema() {
+    // Map jackson node data type to type understood by json
+    classToDataType.put(com.fasterxml.jackson.databind.node.DoubleNode.class.getName(), NUMBER);
+    classToDataType.put(com.fasterxml.jackson.databind.node.TextNode.class.getName(), STRING);
+    classToDataType.put(com.fasterxml.jackson.databind.node.BigIntegerNode.class.getName(),
+        NUMBER);
+    classToDataType.put(com.fasterxml.jackson.databind.node.IntNode.class.getName(), NUMBER);
+    classToDataType.put(com.fasterxml.jackson.databind.node.LongNode.class.getName(), NUMBER);
+    classToDataType.put(com.fasterxml.jackson.databind.node.BooleanNode.class.getName(),
+        BOOLEAN);
+    classToDataType.put(com.fasterxml.jackson.databind.node.NullNode.class.getName(), NULL);
+    classToDataType.put(com.fasterxml.jackson.databind.node.MissingNode.class.getName(), NULL);
+  }
+
+  @VisibleForTesting
+  protected ArrayNode sortJsonArrayList(ArrayNode node) {
+    List<JsonNode> dataNodes = DeriveSchemaUtils.getListFromArray(node);
+    // Sort items of arrayNode using type as the comparator
+    List<JsonNode> sortedDataNodes = dataNodes.stream().distinct()
+        .sorted(Comparator.comparing(o -> o.get("type").asText()))
+        .collect(Collectors.toList());
+    return mapper.createObjectNode().arrayNode().addAll(sortedDataNodes);
+  }
+
+  protected ObjectNode mergeMultipleDataTypes(ObjectNode mergedArray,
+                                              List<JsonNode> primitives,
+                                              List<JsonNode> records,
+                                              List<JsonNode> arrays,
+                                              boolean check2dArray) {
+    ArrayNode items = mapper.createArrayNode();
+    // Adding primitive types to items' list
+    for (JsonNode item : primitives) {
+      items.add(item);
+    }
+    // Merge records if there is at least 1 record
+    if (records.size() > 0) {
```

Comment:
Will there be a use case of calling `mergeRecords` with empty list of records and expecting it to do something? If not, then this check could be delegated moved inside `mergeRecords`. Future callers of this method need not remember to do this check before calling it.

Same applies to `mergeArrays` below

Reply:
The `mergeRecords` function returns an `ObjectNode`. If we pass empty records, it returns `{"type":"object","properties":{}}`. If we remove the check `if (records.size() > 0)` we would have to add another check like `if (mergedRecords.get("properties").size() > 0)` then add it to the `items` list.